[Markdown] fix list punctuation

keith-hall opened the pull request:
Before this change, list punctuation was only scoped for the first item in the list. Now, all list punctuation will be correctly scoped.
Fixes #264

FichteFoll commented on Markdown/Markdown.sublime-syntax:
```diff
@@ -8,14 +8,25 @@ file_extensions:
 scope: text.html.markdown
+variables:
+    separator_line: |-
+        (?x)
```
I suggest always wrapping variables in non-capturing groups and multi-line expressions in `(?x:...)` groups (unless they represent a set). That way you can include them anywhere without having to worry about the including pattern being in extended more or not.

For example, in `contexts.main[1].match` you are already using extended mode, but if you were not the result could be different and unexpected, since extended mode would be active after you included this variable.
However, in `context.separator[0].match` you do not initially use extended mode, but if you were to use any spaces after the inclusion of this variable it would be ignored.

The reason why I wrap everything in a group is because it allows to treat variables atomicly. You can add quantifiers after variables and they mean exactly what you think they would (what the variable matches can be repeated) instead of potentially only repeating the last token of the variable's contents. For example, `{{block_quote}}+` (which makes no sense, but for the example's sake) would expand to `[ ]{,3}.+` and not `(?:[ ]{,3}>.)+`.

keith-hall replied:
that's great advice, thanks @FichteFoll - I'll update my variables :)

keith-hall commented:
I've fixed a few bugs mentioned at https://github.com/sublimehq/Packages/issues/266, added more tests, and started reworking the regex patterns that cause ST to use Oniguruma (https://github.com/sublimehq/Packages/issues/481). I think further improvements should be part of another PR now, as they start to get more complicated (the "italic" and "bold" regexes look intimidating, plus the lookbehinds and `\G`s might be more involved to remove).

wbond commented:
Awesome!

jrappen commented:
Thank you!
